### DISCUSSION TITLE
Is it possible to load entrypoints in Gemfiles from main application?

### DISCUSSION THREAD

**Discussion by vtt** [ElMassimo/vite_ruby#159] [Q&A] [answered]
Supposed I have the following:
### gem1:
```
app/frontend:
  ├── entrypoints:
  │   │── gem1.js 
  │── components:
  │   │ ── # other files
  │── stylesheets:
  │   │── gem1.css
  │── images:
      │── gem1.jpg
```
### gem2:
```
app/frontend:
  ├── entrypoints:
  │   │── gem2.js 
  │── components:
  │   │ ── # other files
  │── stylesheets:
  │   │── gem2.css
  │── images:
      │── gem2.jpg
```
Is it possible to load the gem1.js and gem2.js entrypoints from main app?
ie: <%= vite_javascript_tag 'gem1' %> and <%= vite_javascript_tag 'gem2' %>

**Comment by ElMassimo** [comment, marked as answer]
Technically, you can always import `gem1.js` from an entrypoint in your application, by using an absolute path that resolves to the gem location in disk ([or an alias](https://github.com/ElMassimo/vite_ruby/commit/8da0f6cd704635af940843518bd8778943a9a4fd)).

If you are talking about Rails engines, [take a look at this example setup](https://github.com/ElMassimo/vite_ruby/discussions/95#discussioncomment-976214) with separate compilation, or this example where the [assets can be referenced directly using aliases](https://github.com/ElMassimo/vite_ruby/commit/8da0f6cd704635af940843518bd8778943a9a4fd).

Could you describe the use case you have in mind? Normally you would package assets in an `npm` package instead of a gem, in which case you can import from the package as usual.

**Reply by vtt** [reply]
I have a shared library that was made into a Gem; I don't want to split the JS parts and host on NPM; instead I want both the frontend (JS) and Ruby codes packaged in 1 Gem for easier maintenance.  

The problem with absolute path would not work because of Gem versioning and unknown paths to the Gems locations on disk.

Perhaps there's a way to add additional entrypoint paths using a Ruby configuration file by calling the Gem::Engine.root.path to get the absolute path of the gems?

**Reply by ElMassimo** [reply]
It would be possible to get the actual path in Ruby using `Gem.loaded_specs`, and then set an environment variable using `ViteRuby.env`.

```ruby
ViteRuby.env['GEM1_ASSETS_PATH'] = Gem.loaded_specs['gem1'].full_gem_path + '/app/frontend'
```

Then, in your `vite.config.ts`, you can define an alias to the path:

```js
const gemAssetsPath = process.env.GEM1_ASSETS_PATH

export default defineConfig({
  resolve: {
    alias: {
      '@gem1': gemAssetsPath,
    },
  },   
  server: {
    fs: {
      allow: [gemAssetsPath],
    },
  },
})
```

and then in your main app code:

```js
import example from '@gem1/stylesheets/gem1.css'
```

Seems like a fragile approach, but give it a try and let me know 😄

**Reply by vtt** [reply]
It does work if I manually set the ENV variable.  The problem is how to set this ENV variable for bin/vite when it does assets:precompile?

Using config/initializer to hack the configuration will not work with `bin/vite build` since it's never touched.

Like you said, this is a fragile approach; currently, vite_rails uses a JSON file for configuration; if it can support a Ruby file for configuration then it'd be much easier.

**Reply by ElMassimo** [reply]
A Ruby configuration file would not be readable from the JS plugin.

Since the assets precompile task depends on `:environment`, initializers should be picked up before the task runs.

If for some reason you wanted to skip the default task, you can always add it in bin/vite itself, or use a different executable.

**Reply by vtt** [reply]
Hacking `bin/vite` is not ideal because `bundle exec vite` or even `vite` from installed gem may be called instead. 

Please correct me if I'm wrong; but I believe that `vite_ruby` is converting the configurations into ENV variables.  Why can't the JS Plugin get the configurations from the ENV variables instead of the `config/vite.json` file?

If `vite_ruby` uses a Ruby configuration file (config/vite.rb), then it would support dynamic configuration instead of a static JSON configuration; this opens a world of possibilities.

**Reply by ElMassimo** [reply]
You already have several options:

- Running `assets:precompile` already loads `environment` and initializers
- You can extend the `vite:build` rake task in a `Rakefile` to depend on `:environment` if you need to run it separately

**Reply by vtt** [reply]
Agreed; but all of those are hacky solutions (fragile per your own comment).  The ruby configuration is just a suggestion that would make `vite_ruby` users be more happy due to the possibilities without the need to hack anything.

I noted the two arrayed variables that can't be set via ENV (`additional_entrypoints` and `watch_additional_paths`); a simple use of delimiter `|` would suffice don't you think?

**Reply by ElMassimo** [reply]
> fragile per your own comment

The fragile part is relying on gem structure + `loaded_specs`.

Using initializers or a `Rakefile` to set `ViteRuby.env` is not fragile, and is something some users are using without issues.

---

Some users are already using `npm run vite` directly, and it works when configuring the supported options.

If there's a solid use case for an optional Ruby config file I might consider it, but this use case is too convoluted to be a part of the vision for the core gems :smiley:

I'd recommend that you evaluate wrapping the code in an npm package:
- If the libraries are local, you don't necessarily need to publish it, and can install it using `file:` with a relative path
- You can use a private registry to publish them if you are already publishing the gems

**Reply by vtt** [reply]
Yeah, that's how it's done currently... everytime the library is updated, I have to push the ruby codes to rubygems and the JS codes to npm.  I hope that I could get away from all that by using `vite_ruby`.

I don't agree with the gem structure + `loaded_specs` as fragile.  This will always return the correct absolute paths of the loaded gems; and with the usage of `resolve alias` in `vite` , any frontend assets in any gems can be located and used accordingly.

How about a compromise?

Simply add one liner to `bin/vite` before the last line:
```
load './config/vite.rb' if File.exists?('./config/vite.rb')
```
😃

**Reply by Matt-Yorkley** [reply]
+1 

I think providing a nice way to pull CSS/JS assets into Vite from both `npm` packages _and_ from gems would be very popular in some circles. It might open up new options for gem maintainers, but mostly for all the Rails devs out there toiling away on legacy apps it would be a godsend.

**Reply by Matt-Yorkley** [reply]
It could provide an alternative way to solve some of the "engines in gems" issues as well without the engine having to also configure/run Vite, build the assets independently, serve them on a different port, etc, no?

**Reply by ElMassimo** [reply]
Providing an alternative to the build process inside engines seems reasonable, given how cumbersome the setup is.

Although it would prevent running the vite executable and get the same results, I think that it's a good argument for adding an optional `config/vite.rb` file that is loaded when the config has been resolved.

It would be an "advanced" feature, and the default installation would not include the file. Users that need it can learn about it in the docs.

**Reply by ElMassimo** [reply]
Created https://github.com/ElMassimo/vite_ruby/issues/162 to track this feature request, I might implement it in the next couple of weeks.

**Reply by Matt-Yorkley** [reply]
Just did a bit of hacking around on this and confirmed it works, but the given gem path also needs to be whitelisted under [server.fs.allow](https://vitejs.dev/config/#server-fs-allow) otherwise Vite will (understandably) refuse to load assets from it;
```js
export default defineConfig({
  resolve: {
    alias: {
      '@gem': process.env.GEM_PATH
    },
  },   
  server: {
    fs: {
     allow: [process.env.GEM_PATH]
    },
  },   
})
```

**Reply by ElMassimo** [reply]
You may also set `server.fs.strict` to `false`, but being explicit is better (and safer).

**Reply by Matt-Yorkley** [reply]
I've just been playing with the new additions from #162. Pretty nice!

Having to manually name each gem in the `vite.rb` file and then manually alias and whitelist them in `vite.config.ts` before being able to import their assets in an entrypoint is a bit cumbersome though. Maybe it's okay for 1 or 2, but if it started to get to 5 or 6 it might turn into a bit of a mess...

What do think about exposing all gem-based assets automatically? Something like this would work:

```rb
# config/vite.rb

# *All* the gems in a hash
ViteRuby.env['GEM_PATHS'] = Gem.loaded_specs.to_h{|name, gem| ["gems/#{name}", gem.full_gem_path] }.to_json
```
```js
# config/vite.config.ts

import { defineConfig } from 'vite'
import FullReload from 'vite-plugin-full-reload'
import RubyPlugin from 'vite-plugin-ruby'
import StimulusHMR from 'vite-plugin-stimulus-hmr'

const gem_paths = JSON.parse(process.env.GEM_PATHS)

export default defineConfig({
  plugins: [
    RubyPlugin(),
    FullReload(['config/routes.rb', 'app/views/**/*'], { delay: 200 }),
    StimulusHMR(),
  ],
  resolve: {
    alias: gem_paths
  },
  server: {
    fs: {
      allow: [
        ...Object.values(gem_paths)
      ]
    }
  }
})
```

```js
# app/frontend/entrypoints/application.js

import "gems/your_favourite_gem/assets/js/whatever_you_want.js"
import "gems/some_css_framework/dist/css/hot_theme.css"
```

With the new bundling options in Rails, it's becoming more normal for modern gems to include assets. https://github.com/hotwired/stimulus-rails/tree/main/app/assets/javascripts is an example. 

What do you think?

**Reply by ElMassimo** [reply]
I'd like to wait and see how frequent this usage is for Vite Ruby users before expanding the API.

In the meantime, users can leverage `config/vite.rb` to implement a pattern like the one you described above.

Thanks for sharing! Feel free to open a new discussion to share this "recipe".

**Reply by mameier** [reply]
Vote up for this "recipe": good to keep all in sync for gems like actiontext, stimulus, ...  
no need to find and install "matching" npm modules.
